les/downloader/statesync.go

```diff
@@ -586,13 +585,35 @@ func (s *stateSync) process(req *stateReq) (int, error) {
 // processNodeData tries to inject a trie node data blob delivered from a remote
 // peer into the state trie, returning whether anything useful was written or any
 // error occurred.
-func (s *stateSync) processNodeData(blob []byte) (common.Hash, error) {
-	res := trie.SyncResult{Data: blob}
+//
+// If multiple requests correspond to the same hash, this method will inject the
+// blob as a result for the first one only, leaving the remaining duplicates to
+// be fetched again.
+func (s *stateSync) processNodeData(nodeTasks map[string]*trieTask, codeTasks map[common.Hash]*codeTask, blob []byte) (common.Hash, error) {
+	var hash common.Hash
 	s.keccak.Reset()
 	s.keccak.Write(blob)
-	s.keccak.Read(res.Hash[:])
-	err := s.sched.Process(res)
-	return res.Hash, err
+	s.keccak.Read(hash[:])
+
+	if _, present := codeTasks[hash]; present {
+		err := s.sched.ProcessCode(trie.CodeSyncResult{
+			Hash: hash,
```

Comment:
In the previous logic. 
```
err := s.sched.Process(res)
```
They will forward res to Process function , in the new logic, we need to split into handle either ProcessCode or ProcessNode. Could u explain a little the condition check here for choosing which kind of processing in sched.

Reply:
Since the requests list are also now divided into `nodeReqs` and `codeReqs`, we use `ProcessNode` for each request taking from `nodeReqs` and `ProcessCode` for each request taking from `codeReqs`